core/editable.js

```diff
@@ -1647,7 +1647,12 @@
 		// guarantee it's result to be a valid DOM tree.
 		function insert( editable, type, data, range ) {
 			var editor = editable.editor,
-				dontFilter = false;
+				dontFilter = false,
+				html = editable.getHtml(),
+				// Instead of getData method, we directly check the HTML
+				// due to the fact that interal getData operates on latest snapshot,
+				// not the current content (#4301).
+				isEmptyEditable = html === '' || html.match( emptyParagraphRegexp );
```

Comment:
Actually `emptyParagraphRegexp` searches for empty `div`, `p` and several other elements. It's not a paragraph as an element, but just as a block of text.

Reply:
Thanks for explaining that, but still we are missing unit and manual tests for that case. I just wanted to add them quickly during this review, but find out that I'm actually not able to create any easily reproducible test case for that scenario. Could you update your PR with these tests for empty container elements? If it's just defensive programming, it would be still good to have some spying tests instead.